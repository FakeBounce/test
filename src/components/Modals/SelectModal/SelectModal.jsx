import React, { Component } from "react";
import { Button } from "reactstrap";
import { observer } from "mobx-react";
import { observable } from "mobx";
import { translate } from "common/methods/translations";

import UserModel from "models/UserModel";
import Modal from "elements/Modal/Modal";
import _ from "lodash";
import validateEmail from "common/methods/ValidateEmail";

@observer
class SelectModal extends Component {
  @observable firstName = "";
  @observable lastName = "";
  @observable email = "";
  @observable function_ = "";
  @observable phoneNumber = "";
  @observable profile = "";

  constructor() {
    super();
    this.handleFirstNameChange = this.handleFirstNameChange.bind(this);
    this.handleLastNameChange = this.handleLastNameChange.bind(this);
    this.handleEmailChange = this.handleEmailChange.bind(this);
    this.handleFunction_Change = this.handleFunction_Change.bind(this);
    this.handlePhoneNumberChange = this.handlePhoneNumberChange.bind(this);
    this.selectedProfile = this.selectedProfile.bind(this);
    this.resetFields = this.resetFields.bind(this);
    this.debounceInput = _.debounce(this.debounceInput, 600);
  }

  resetFields() {
    this.firstName = null;
    this.lastName = null;
    this.email = null;
    this.function_ = null;
    this.phoneNumber = null;
    this.profile = null;
  }

  saveButtonClick() {
    const { handleSaveButtonClick } = this.props;
    const newUser = new UserModel(
      null,
      this.firstName,
      this.lastName,
      this.email,
      this.function_,
      this.phoneNumber,
      this.profile
    );
    this.resetFields();
    handleSaveButtonClick(newUser);
  }

  saveButton() {
    return (
      <Button color="primary" onClick={this.saveButtonClick.bind(this)}>
        {translate("globals.save")}
      </Button>
    );
  }

  debounceInput(email) {
    if (validateEmail(email)) {
      if (email) {
        //TODO check if email exist if so throw appropriate message
        console.log("email can be used");
      } else {
        console.log("email already exist");
      }
    } else {
      //TODO throw wrong email format
      console.log("wrong email format");
    }
  }

  handleEmailChange(email) {
    this.email = email;
    this.debounceInput(email);
  }

  handleFunction_Change(function_) {
    this.function_ = function_;
  }

  handlePhoneNumberChange(phoneNumber) {
    this.phoneNumber = phoneNumber;
  }

  selectedProfile(profile) {
    this.profile = profile;
  }

  handleFirstNameChange(firstName) {
    this.firstName = firstName;
  }

  handleLastNameChange(lastName) {
    this.lastName = lastName;
  }

  render() {
    return (
      <Modal
        ModalFooterContent={this.saveButton.bind(this)}
        {...this.props}
        className={"FormModal"}
      >
        <p>{translate("myProfile.alerts.contractSelection")}</p>
      </Modal>
    );
  }
}

export default SelectModal;
