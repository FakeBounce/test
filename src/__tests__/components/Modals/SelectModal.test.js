import React from "react";
import { shallow, configure } from "enzyme/build";
import Adapter from "enzyme-adapter-react-16/build";
import SelectModal from "components/Modals/SelectModal/SelectModal";

configure({ adapter: new Adapter() });

const handleSave = jest.fn();

const defaultTestingProps = {
  isOpen: false,
  title: "Title",
  handleSaveButtonClick: handleSave,
  toggle: () => jest.fn()
};

describe("Testing SelectModal component", () => {
  it("renders as expected without parameters", () => {
    const wrapper = shallow(<SelectModal />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders as expected with parameters (as closed)", () => {
    const wrapper = shallow(<SelectModal {...defaultTestingProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders as expected with parameters (as open)", () => {
    const wrapper = shallow(<SelectModal {...defaultTestingProps} isOpen />);
    expect(wrapper).toMatchSnapshot();
  });

  it("behave correctly when calling props", () => {
    const wrapper = shallow(<SelectModal {...defaultTestingProps} />);

    wrapper.instance().saveButtonClick();
    expect(handleSave.mock.calls.length).toBe(1);
    // expect(handleYes.mock.calls.length).toBeCalledWith(false);
  });

  it("behave correctly when props are invalid", () => {
    const wrapper = shallow(
      <SelectModal {...defaultTestingProps} handleSaveButtonClick={true} />
    );
    try {
      wrapper.instance().saveButtonClick();
      // Fail test if above expression doesn't throw anything.
      expect(true).toBe(false);
    } catch (e) {
      expect(e).toBeInstanceOf(TypeError);
      expect(e.message).toBe("handleSaveButtonClick is not a function");
    }
  });

  // @TODO Check if functions are used and test'em
});