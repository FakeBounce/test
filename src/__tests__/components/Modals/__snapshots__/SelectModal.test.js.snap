// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SelectModal component renders as expected with parameters (as closed) 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectModal
    handleSaveButtonClick={[Function]}
    isOpen={false}
    title="Title"
    toggle={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "ModalFooterContent": [Function],
      "children": <p />,
      "className": "FormModal",
      "handleSaveButtonClick": [Function],
      "isOpen": false,
      "title": "Title",
      "toggle": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
      },
      "ref": null,
      "rendered": "",
      "type": "p",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "ModalFooterContent": [Function],
        "children": <p />,
        "className": "FormModal",
        "handleSaveButtonClick": [Function],
        "isOpen": false,
        "title": "Title",
        "toggle": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Array [],
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Testing SelectModal component renders as expected with parameters (as open) 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectModal
    handleSaveButtonClick={[Function]}
    isOpen={true}
    title="Title"
    toggle={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "ModalFooterContent": [Function],
      "children": <p />,
      "className": "FormModal",
      "handleSaveButtonClick": [Function],
      "isOpen": true,
      "title": "Title",
      "toggle": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
      },
      "ref": null,
      "rendered": "",
      "type": "p",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "ModalFooterContent": [Function],
        "children": <p />,
        "className": "FormModal",
        "handleSaveButtonClick": [Function],
        "isOpen": true,
        "title": "Title",
        "toggle": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Array [],
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Testing SelectModal component renders as expected without parameters 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectModal />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "ModalFooterContent": [Function],
      "children": <p />,
      "className": "FormModal",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
      },
      "ref": null,
      "rendered": "",
      "type": "p",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "ModalFooterContent": [Function],
        "children": <p />,
        "className": "FormModal",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Array [],
  Symbol(enzyme.__childContext__): null,
}
`;
